spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -309,30 +346,45 @@ public void deliveryComplete(IMqttToken token) {
 
 	@Override
 	public void connectComplete(boolean reconnect, String serverURI) {
-		if (!reconnect) {
-			ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
-			String[] topics = getTopic();
-			this.topicLock.lock();
-			try {
-				if (topics.length > 0) {
-					int[] requestedQos = getQos();
-					this.mqttClient.subscribe(topics, requestedQos).waitForCompletion(getCompletionTimeout());
-					String message = "Connected and subscribed to " + Arrays.toString(topics);
-					logger.debug(message);
-					if (applicationEventPublisher != null) {
-						applicationEventPublisher.publishEvent(new MqttSubscribedEvent(this, message));
-					}
-				}
+		if (reconnect) {
+			return;
+		}
+		var clientManager = getClientManager();
+		if (clientManager != null && this.mqttClient == null) {
+			this.mqttClient = clientManager.getClient();
+		}
+
+		String[] topics = getTopic();
+		ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
+		this.topicLock.lock();
+		try {
+			if (topics.length == 0) {
+				return;
 			}
-			catch (MqttException ex) {
-				if (applicationEventPublisher != null) {
-					applicationEventPublisher.publishEvent(new MqttConnectionFailedEvent(this, ex));
-				}
-				logger.error(ex, () -> "Error subscribing to " + Arrays.toString(topics));
+
+			int[] requestedQos = getQos();
+			MqttSubscription[] subscriptions = IntStream.range(0, topics.length)
+					.mapToObj(i -> new MqttSubscription(topics[i], requestedQos[i]))
+					.toArray(MqttSubscription[]::new);
+			MessageListener[] listeners = IntStream.range(0, topics.length)
+					.mapToObj(t -> new MessageListener())
```

Comment:
I think we can hold a single instance for this `IMqttMessageListener` per whole subscription, at least as a variable for all items of this array.